official/recommendation/data_pipeline.py

```diff
@@ -0,0 +1,727 @@
+    map_args = [i for i in range(self.eval_batches_per_epoch)]
+    with popen_helper.get_threadpool(6) as pool:
+      pool.map(self._get_eval_batch, map_args)
+    self._eval_dataset.end_construction()
+
+    tf.logging.info("Eval construction complete. Time: {:.1f} seconds".format(
+        timeit.default_timer() - start_time))
+
+  def make_input_fn(self, is_training):
+    if self._fatal_exception is not None:
+      raise ValueError("Fatal exception in the data production loop: {}"
+                       .format(self._fatal_exception))
+
+    return (
+        self._train_dataset.make_input_fn(self.train_batch_size) if is_training
+        else self._eval_dataset.make_input_fn(self.eval_batch_size))
+
+
+class DummyConstructor(threading.Thread):
+  """Class for running with synthetic data."""
+  def run(self):
+    pass
+
+  def stop_loop(self):
+    pass
+
+  @staticmethod
+  def make_input_fn(is_training):
+    """Construct training input_fn that uses synthetic data."""
+
+    def input_fn(params):
+      """Generated input_fn for the given epoch."""
+      batch_size = (params["batch_size"] if is_training else
+                    params["eval_batch_size"] or params["batch_size"])
+      num_users = params["num_users"]
+      num_items = params["num_items"]
+
+      users = tf.random_uniform([batch_size], dtype=tf.int32, minval=0,
+                                maxval=num_users)
+      items = tf.random_uniform([batch_size], dtype=tf.int32, minval=0,
+                                maxval=num_items)
+
+      if is_training:
+        valid_point_mask = tf.cast(tf.random_uniform(
+            [batch_size], dtype=tf.int32, minval=0, maxval=2), tf.bool)
+        labels = tf.cast(tf.random_uniform(
+            [batch_size], dtype=tf.int32, minval=0, maxval=2), tf.bool)
+        data = {
+            movielens.USER_COLUMN: users,
+            movielens.ITEM_COLUMN: items,
+            rconst.VALID_POINT_MASK: valid_point_mask,
+        }, labels
+      else:
+        dupe_mask = tf.cast(tf.random_uniform([batch_size], dtype=tf.int32,
+                                              minval=0, maxval=2), tf.bool)
+        data = {
+            movielens.USER_COLUMN: users,
+            movielens.ITEM_COLUMN: items,
+            rconst.DUPLICATE_MASK: dupe_mask,
+        }
+
+      dataset = tf.data.Dataset.from_tensors(data).repeat(
+          rconst.SYNTHETIC_BATCHES_PER_EPOCH * params["batches_per_step"])
+      dataset = dataset.prefetch(32)
+      return dataset
+
+    return input_fn
+
+
+class MaterializedDataConstructor(BaseDataConstructor):
+  """Materialize a table of negative examples for fast negative generation.
+
+  This class creates a table (num_users x num_items) containing all of the
+  negative examples for each user. This table is conceptually ragged; that is to
+  say the items dimension will have elements at the end which are not used equal
+  to the number of positive elements for a given user. For instance:
```

Comment:
> that is to say the items dimension will have elements at the end which are not used equal to the number of positive elements for a given user

I don't understand that clause. Removing the word "used" makes it grammatically correct